[IC3 ACS - Auth] <preview/2025-03-02-preview> Durable identity support for BYOI, ACS Token Exchange for Teams Phone Extensibility and for Entra  user support

stefang931 opened the pull request:
# Data Plane API - Pull Request
This pull request introduces support for reusing identities, which will unlock higher business value for developers. By reusing identities, developers can link multiple usage sessions to the same underlying user, providing a more cohesive and personalized user experience.

Currently, developers face challenges in reusing identities because they are required to maintain their own database to store mappings between their users and the Azure Communication Services (ACS) identities they create. This additional burden complicates the development process and can deter developers from adopting identity reuse practices.

With the new support for reusing identities, we aim to simplify this process, making it easier for developers to implement identity reuse. This enhancement will encourage more developers to reuse identities, ultimately benefiting both the developers and the business.

## API Info: The Basics
Most of the information about your service should be captured in the issue that serves as your [*API Spec engagement record*](https://aka.ms/azsdk/onboarding/restapischedule).

* Link to API Spec engagement record issue:

Is this review for (select one):

- [ ] a private preview
- [X] a public preview
- [ ] GA release

### Change Scope

<sup>This section will help us focus on the specific parts of your API that are new or have been modified. <br/>Please share a link to the design document for the new APIs, a link to the previous API Spec document (if applicable), and the root paths that have been updated. </sup>
* Design Document: https://skype.visualstudio.com/SPOOL/_wiki/wikis/SPOOL.wiki/50705/-Proposal-Durable-identity-for-BYOI
* Previous API Spec Doc: [Stable <2023-10-01>](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/communication/data-plane/Identity/stable/2023-10-01/CommunicationIdentity.json)
* Updated paths: TODO

### Viewing API changes

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

### Suppressing failures

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[Swagger-Suppression-Process](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/85/Swagger-Suppression-Process) 
to get approval.

## ❔Got questions? Need additional info?? We are here to help!

<details>
  <summary> Contact us!</summary>

The [Azure API Review Board](https://aka.ms/azsdk/onboarding/restapischedule) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azsdk/onboarding/restapischedule).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling

 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://aka.ms/style)
 * [Open API Hub](https://aka.ms/openapiportal)

### Guidelines & Specifications

 * [Azure REST API Guidelines](https://aka.ms/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links

 * [Azure DevTools Wiki](https://aka.ms/azapi)

</details>

fix https://github.com/Azure/azure-rest-api-specs/issues/26005


# Teams Phone Extensibility

## Examples

**As Communication Service resource owner, assigning Teams Phone user to ACS resource.**
For Teams Phone user with oid _objectId_ from tenant _tenantId_. ACS Resource is defined by request host.
Assignment created this way allows the Microsoft Teams user with oid _objectId_ from tenant _tenantId_ to use Entra application _appId_ to get an access token for using Teams Phone extensibility with ACS resource.

```
async Task CreateTeamsExtensionAssignmentExample(ObjectId objectId, ObjectId appId, TenantId tenantId, CancellationToken cancellationToken)
{
    var assignmentRequest = new TeamsExtensionAssignmentCreateOrUpdateRequest(TeamsPhonePrincipalType.User, new[] { appId });
    
    var json = JsonConvert.SerializeObject(assignmentRequest);
    var httpRequest = new HttpRequestMessage(HttpMethod.Put, $"access/teamsExtension/tenants/{tenantId}/assignments/{objectId}?api-version=2025-03-02-preview")
    {
        Content = new StringContent(json, Encoding.UTF8, "application/json"),
    };
    await AddHmacAuthenticationAsync(httpRequest);
    
    var assignmentResponse = await SendAsync<TeamsExtensionAssignmentResponse>(httpRequest, cancellationToken);
    var assignment = assignmentResponse.Value;
    
    Console.WriteLine($"Assignment principal type: {assignment.PrincipalType}");
    Console.WriteLine($"Assignment tenant id: {assignment.TenantId}");
    Console.WriteLine($"Assignment object id: {assignment.ObjectId}");
    Console.WriteLine($"Assignment client ids: {string.Join(", ", assignment.ClientIds)}");
}
```

**As Communication Service resource owner, getting existing assignment of the Teams Phone user to ACS resource.**
For Teams Phone user with oid _objectId_ from tenant _tenantId_. ACS Resource is defined by request host.

```
async Task GetTeamsExtensionAssignmentExample(TenantId tenantId, ObjectId objectId, CancellationToken cancellationToken)
{
    var httpRequest = new HttpRequestMessage(HttpMethod.Get, $"access/teamsExtension/tenants/{tenantId}/assignments/{objectId}?api-version=2025-03-02-preview");
    await AddHmacAuthenticationAsync(httpRequest);

    var assignmentResponse = await SendAsync<TeamsExtensionAssignmentResponse>(httpRequest, cancellationToken);
    var assignment = assignmentResponse.Value;

    Console.WriteLine($"Assignment principal type: {assignment.PrincipalType}");
    Console.WriteLine($"Assignment tenant id: {assignment.TenantId}");
    Console.WriteLine($"Assignment object id: {assignment.ObjectId}");
    Console.WriteLine($"Assignment client ids: {string.Join(", ", assignment.ClientIds)}");
}
```
**As Communication Service resource owner, deleting existing assignment of the Teams Phone user to ACS resource.**
For Teams Phone user with oid _objectId_ from tenant _tenantId_. ACS Resource is defined by request host.
With assignment deleted, Teams Phone user can't get token for using Teams Phone extensibility with ACS resource any more and existing tokens would be revoked.

```
async Task DeleteTeamsExtensionAssignmentExample(TenantId tenantId, ObjectId objectId, CancellationToken cancellationToken)
{
    var httpRequest = new HttpRequestMessage(HttpMethod.Delete, $"access/teamsExtension/tenants/{tenantId}/assignments/{objectId}?api-version=2025-03-02-preview");
    await AddHmacAuthenticationAsync(httpRequest);

    var response = await SendAsync<HttpResponseMessage>(httpRequest, cancellationToken);

    if (response.IsSuccessStatusCode)
    {
        Console.WriteLine("Assignment deleted successfully.");
    }
    else
    {
        Console.WriteLine($"Failed to delete assignment. Status code: {response.StatusCode}");
    }
}
```

**As Teams Phone user, use Entra User Access token to get access token for using Teams Phone extensibility with ACS resource**
ACS Resource is defined by request host.
This call won't be successful until ACS resource owner creates the assignment for the Teams Phone user in the ACS resource.

```
async Task ExchangeTeamsPhoneTokenExample(CancellationToken cancellationToken)
{
    var entraAccessToken = GetEntraAccessTokenAsync();
    var exchangeTokenRequest = new TeamsExtensionExchangeTokenRequest();

    var json = JsonConvert.SerializeObject(exchangeTokenRequest);
    var httpRequest = new HttpRequestMessage(HttpMethod.Post, $"access/teamsExtension/:exchangeAccessToken?api-version=2025-03-02-preview")
    {
        Content = new StringContent(json, Encoding.UTF8, "application/json"),
    };

    httpRequest.Headers.Add(HeaderNames.Authorization, $"Bearer {entraAccessToken}");

    var response = await SendAsync<CommunicationIdentityAccessTokenResult>(httpRequest, cancellationToken);

    if (response.IsSuccessStatusCode)
    {
        Console.WriteLine($"Identity: {response.Value.Identity}");
        Console.WriteLine($"Access token: {response.Value.AccessToken}");
    }
    else
    {
        Console.WriteLine($"Failed to exchange token. Status code: {response.StatusCode}");
    }
}
```

## Core Concepts

**Glossary**
*Teams Phone* - telephony service integrated with Microsoft Teams
*ACS* - Azure Communication Services

**Teams Phone extensibility** - the integration of Teams Phone and ACS, allowing the developers to leverage Teams Phone capabilities in their ACS based applications.
**Assignment of Teams Phone user** - allowing the Teams Phone users to use Teams Phone extensibility with defined ACS resource. Any single user can be assigned to any number of ACS resources. Only ACS resource owner can create or delete the assignment.
**Token exchange** - ACS doesn't accept the Entra access token of Teams Phone user directly. It is necessary to do the token exchange that takes the Teams Phone user token and returns the token usable for Teams Phone extension to work with specified ACS resource. 
*Note*: Calling of the token exchange would be wrapped in the SDK in most cases and the Teams Phone user doesn't need to care about it.


# Entra user support

**As Communication Service resource owner, assigning Entra principal to ACS resource.**
For Entra principal (user in the sample) with oid _objectId_ from tenant _tenantId_. ACS Resource is defined by request host.
Assignment created this way allows the Microsoft Teams user with oid _objectId_ from tenant _tenantId_ to use Entra application _appId_ to get an access token for using ACS data plane in the context of ACS resource.

```
async Task CreateEntraAssignmentExample(ObjectId objectId, ObjectId appId, TenantId tenantId, CancellationToken cancellationToken)
{
    var assignmentRequest = new EntraAssignmentCreateOrUpdateRequest(tenantId, EntraPrincipalType.User, new[] { appId });

    var json = JsonConvert.SerializeObject(assignmentRequest);
    using var httpRequest = new HttpRequestMessage(HttpMethod.Put, $"access/entra/assignments/{objectId}?api-version=2025-03-02-preview")
    {
        Content = new StringContent(json, Encoding.UTF8, "application/json"),
    };
    await AddHmacAuthenticationAsync(httpRequest);

    var assignmentResponse = await SendAsync<EntraAssignment>(httpRequest, cancellationToken);
    var assignment = assignmentResponse.Value;

    Console.WriteLine($"Assignment principal type: {assignment.PrincipalType}");
    Console.WriteLine($"Assignment tenant id: {assignment.TenantId}");
    Console.WriteLine($"Assignment object id: {assignment.ObjectId}");
    Console.WriteLine($"Assignment client ids: {string.Join(", ", assignment.ClientIds)}");
}
```

**As Communication Service resource owner, getting existing assignment of the Entra principal to ACS resource.**
For Entra principal with oid _objectId_. ACS Resource is defined by request host.

```
async Task GetEntraAssignmentExample(ObjectId objectId, CancellationToken cancellationToken)
{
    using var httpRequest = new HttpRequestMessage(HttpMethod.Get, $"access/entra/assignments/{objectId}?api-version=2025-03-02-preview");
    await AddHmacAuthenticationAsync(httpRequest);

    var assignmentResponse = await SendAsync<EntraAssignment>(httpRequest, cancellationToken);
    var assignment = assignmentResponse.Value;

    Console.WriteLine($"Assignment principal type: {assignment.PrincipalType}");
    Console.WriteLine($"Assignment tenant id: {assignment.TenantId}");
    Console.WriteLine($"Assignment object id: {assignment.ObjectId}");
    Console.WriteLine($"Assignment client ids: {string.Join(", ", assignment.ClientIds)}");
}
```

**As Communication Service resource owner, deleting existing assignment of the Entra principal to ACS resource.**
For Entra user with oid _objectId_. ACS Resource is defined by request host.
With assignment deleted, Entra user can't get token for using ACS data-plane in ACS resource context any more and existing tokens would be revoked.

```
async Task DeleteEntraAssignmentExample(ObjectId objectId, CancellationToken cancellationToken)
{
    using var httpRequest = new HttpRequestMessage(HttpMethod.Delete, $"access/entra/assignments/{objectId}?api-version=2025-03-02-preview");
    await AddHmacAuthenticationAsync(httpRequest);

    var response = await SendAsync<HttpResponseMessage>(httpRequest, cancellationToken);

    if (response.IsSuccessStatusCode)
    {
        Console.WriteLine("Assignment deleted successfully.");
    }
    else
    {
        Console.WriteLine($"Failed to delete assignment. Status code: {response.StatusCode}");
    }
}
```

**As Communication Service resource owner, list all existing assignment of the Entra users to ACS resource.**
ACS Resource is defined by request host. Paged request.

```
async Task ListEntraAssignmentsExample(string nextLink, CancellationToken cancellationToken)
{
    using var httpRequest = new HttpRequestMessage(HttpMethod.Get, $"access/entra/assignments?api-version=2025-03-02-preview&nextLink={nextLink}");
    await AddHmacAuthenticationAsync(httpRequest);
    var assignmentResponse = await SendAsync<EntraAssignmentsResponse>(httpRequest, cancellationToken);
    var assignments = assignmentResponse.Value.Value;
    foreach (var assignment in assignments)
    {
        Console.WriteLine($"Assignment principal type: {assignment.PrincipalType}");
        Console.WriteLine($"Assignment tenant id: {assignment.TenantId}");
        Console.WriteLine($"Assignment object id: {assignment.ObjectId}");
        Console.WriteLine($"Assignment client ids: {string.Join(", ", assignment.ClientIds)}");
    }
}
```

**As Communication Service resource owner, batch update of assignments of the Entra users to ACS resource.**
ACS Resource is defined by request host.
It is possible to upsert batch of users and delete batch of users in single request.

```
async Task UpdateEntraAssignmentsExample(IList<EntraAssignment> toUpsert, IList<EntraAssignment> toDelete, CancellationToken cancellationToken)
{
    var request = new EntraAssignmentsPatchRequest();
    foreach (var item in toUpsert)
    {
        var assignmentRequest = new EntraAssignmentCreateOrUpdateRequest(item.TenantId, item.PrincipalType, item.ClientIds);
        request.Add(item.ObjectId, assignmentRequest);
    }

    foreach (var item in toDelete)
    {
        request.Add(item.ObjectId, null);
    }

    var json = JsonConvert.SerializeObject(request);
    using var httpRequest = new HttpRequestMessage(new HttpMethod("PATCH"), $"access/entra/assignments?api-version=2025-03-02-preview")
    {
        Content = new StringContent(json, Encoding.UTF8, "application/json"),
    };
    var response = await SendAsync<EntraAssignmentsResponse>(httpRequest, cancellationToken);
    var assignments = response.Value.Value;
    foreach (var assignment in assignments)
    {
        Console.WriteLine($"Assignment principal type: {assignment.PrincipalType}");
        Console.WriteLine($"Assignment tenant id: {assignment.TenantId}");
        Console.WriteLine($"Assignment object id: {assignment.ObjectId}");
        Console.WriteLine($"Assignment client ids: {string.Join(", ", assignment.ClientIds)}");
    }
}
```

**As an Entra user, use User Access token to get access token for using ACS data-plane capabilities in the context of ACS resource**
ACS Resource is defined by request host.
This call won't be successful until ACS resource owner creates the assignment for the Entra user in the ACS resource.

```
internal async Task ExchangEntraTokenExample(ObjectId objectId, ObjectId appId, CancellationToken cancellationToken)
{
    var entraAccessToken = GetEntraAccessTokenAsync();

    var exchangeTokenRequest = new EntraExchangeTokenRequest();

    var json = JsonConvert.SerializeObject(exchangeTokenRequest);
    using var httpRequest = new HttpRequestMessage(HttpMethod.Post, $"access/entra/:exchangeAccessToken?api-version=2025-03-02-preview")
    {
        Content = new StringContent(json, Encoding.UTF8, "application/json"),
    };

    httpRequest.Headers.Add(HeaderNames.Authorization, $"Bearer {entraAccessToken}");

    var response = await SendAsync<CommunicationIdentityAccessTokenResult>(httpRequest, cancellationToken);

    if (response.IsSuccessStatusCode)
    {
        Console.WriteLine($"Identity: {response.Value.Identity}");
        Console.WriteLine($"Access token: {response.Value.AccessToken}");
    }
    else
    {
        Console.WriteLine($"Failed to exchange token. Status code: {response.StatusCode}");
    }
}
```

## Core Concepts

**Entra principal** - Entra entity that can be assigned to ACS resource. It can be either Entra user, whole tenant or single group. Both Entra workforce and Entra external are possible.

**Assignment of Entra principal** - allowing the Entra user, group of Entra users or all users of Entra tenant to use ACS in de context of defined ACS resource. Any single principal can be assigned to any number of ACS resources. Only ACS resource owner can create or delete the assignment.
**Token exchange** - ACS doesn't accept the user access tokens of Entra users directly. It is necessary to do the token exchange that takes the Entra user token and returns the token usable for ACS data-plane.

DominikMe commented on specification/communication/data-plane/Identity/preview/2025-04-01-preview/CommunicationIdentity.json:
```diff
@@ -328,6 +383,27 @@
+        },
+        "externalId": {
+          "description": "The external Id if one has been associated with the identity.",
+          "type": "string"
```
Should we add minLength 1 and maxLength 256?

DominikMe commented on specification/communication/data-plane/Identity/preview/2025-04-01-preview/CommunicationIdentity.json:
```diff
@@ -311,6 +362,10 @@
     "CommunicationIdentityCreateRequest": {
       "type": "object",
       "properties": {
+        "externalId": {
```
Should we add minLength 1 and maxLength 256?

stefang931 replied:
Done

DominikMe commented on specification/communication/data-plane/Identity/preview/2025-04-01-preview/CommunicationIdentity.json:
```diff
@@ -0,0 +1,501 @@
+  "info": {
+    "title": "CommunicationIdentityClient",
+    "description": "Azure Communication Identity Service",
+    "version": "2023-10-01"
```
Change to 

```suggestion
    "version": "2025-04-01-preview"
```

DominikMe commented on specification/communication/data-plane/Identity/preview/2025-04-01-preview/examples/DeleteIdentity.json:
```diff
@@ -0,0 +1,11 @@
+{
+  "parameters": {
+    "api-version": "2023-10-01",
```
```suggestion
    "api-version": "2025-04-01-preview",
```

DominikMe commented on specification/communication/data-plane/Identity/preview/2025-04-01-preview/examples/GetIdentity.json:
```diff
@@ -0,0 +1,17 @@
+{
+  "parameters": {
+    "api-version": "2023-10-01",
```
```suggestion
    "api-version": "2025-04-01-preview",
```

DominikMe commented on specification/communication/data-plane/Identity/preview/2025-04-01-preview/examples/IssueAccessToken.json:
```diff
@@ -0,0 +1,26 @@
+{
+  "parameters": {
+    "api-version": "2023-10-01",
```
```suggestion
    "api-version": "2025-04-01-preview",
```

DominikMe commented on specification/communication/data-plane/Identity/preview/2025-04-01-preview/examples/RevokeAccessTokens.json:
```diff
@@ -0,0 +1,11 @@
+{
+  "parameters": {
+    "api-version": "2023-10-01",
```
```suggestion
    "api-version": "2025-04-01-preview",
```

DominikMe commented on specification/communication/data-plane/Identity/preview/2025-04-01-preview/examples/CreateIdentity.json:
```diff
@@ -0,0 +1,44 @@
+  "responses": {
+    "201": {
+      "body": {
+        "identity": {
```
I'm wondering if we should return `lastTokenIssuedAt` here as well, then we'd have `identity` be the same schema as on the GET route. But it would tie them together and GET identity will maybe later return other Entra properties that aren't relevant to BYOI. Could still be okay as they will be optional.

Commit:
Baseline commit

Commit:
Add API changes

cemateia committed:
Add the OPS token exchange to the swagger (#33127)

* Add the OPS token exchange to the swagger

* run prettier

* add access APIs

* run prettifier

* revert changes

* fix

* Update CommunicationIdentity.json

Changing OPS API route according to private review

* Fixing errors in swagger

Errors found in swagger analysis

* Fixing another errors in swagger

Fixing issues in swagger

* Update ExchangeTeamsPhoneToken.json

Typo

* Update GetTeamsPhoneAssignment.json

typo

* Update CommunicationIdentity.json

Principaltype as definition to calm down swagger linter

* Update CreateOrReplaceTeamsPhoneAssignment.json

prettier

* Update CommunicationIdentity.json

Fixing location of principal type in swagger

* Update CommunicationIdentity.json

updated according to internal API review

* Update CommunicationIdentity.json

Small updates in descriptions

* Apply suggestions from code review

fix teams principal type in examples

---------

Co-authored-by: paveldostalms <paveldostal@microsoft.com>
Co-authored-by: Dominik <domessin@microsoft.com>

paveldostalms committed:
Update CommunicationIdentity.json

Added Entra assignment

paveldostalms committed:
Examples for Entra assignment

Added examples for Entra assignment

paveldostalms committed:
Add Entra token exchange

Adding entra token exchange API + example

Commit:
Fix GetIdentity sample

paveldostalms committed:
User principal in entraid assignment samples

Using user instead of resourceAccount in samples. Fixing some issues from PR checks

paveldostalms committed:
Update UpdateEntraAssignments.json

Fixing sample

paveldostalms committed:
Update CreateOrReplaceEntraAssignment.json

Typo

paveldostalms committed:
Examples prettier

paveldostalms committed:
Updating operationId

changes according to linter requirements

paveldostalms committed:
Update CommunicationIdentity.json

Prettier, According to linter errors

paveldostalms committed:
Using TeamsExtension instead of TeamsPhone

Renaming structures for access/teamsExtension

paveldostalms committed:
Update CommunicationIdentity.json

Forgotten rename

paveldostalms committed:
Update CommunicationIdentity.json

Typo

DominikMe committed:
remove teamsResourceAccount from Entra routes, and make clientIds required

DominikMe committed:
don't use an anonnymous type for patch update

DominikMe committed:
avoid term upsert to fix spell check

DominikMe committed:
swagger fixes and TeamsPhone rename

DominikMe committed:
add x-ms-pageable

DominikMe committed:
Merge branch 'main' into gistefan/externalId_apis

DominikMe committed:
satisfy breaking changes check

DominikMe committed:
Merge branch 'gistefan/externalId_apis' of https://github.com/Azure/azure-rest-api-specs into gistefan/externalId_apis

DominikMe committed:
delete accidentally commited output file

DominikMe committed:
address Stewardship board feedback

DominikMe committed:
update renamed property in example files

DominikMe committed:
fix returned status code to 201 in both swagger and examples

stefang931 committed:
Merge branch 'main' into gistefan/externalId_apis

weidongxu-microsoft reviewed (approved):
LGTM
